File: langs/java/linpack/Linpack.java
package linpack;

import java.util.Arrays;
import java.util.Random;

public class Linpack {

    public static final int ARRAY_SIZE = 2000;

    public static class LinpackResult {
        public double norma;
        public double residual;
        public double normalisedResidual;
        public double epsilon;
        public double time;
        public double mflops;

        @Override
        public String toString() {
            return "LinpackResult{" +
                    "norma=" + norma +
                    ", residual=" + residual +
                    ", normalisedResidual=" + normalisedResidual +
                    ", epsilon=" + epsilon +
                    ", time=" + time +
                    ", mflops=" + mflops +
                    '}';
        }
    }

    public static long second() {
        return System.nanoTime() / 1000000;
    }

    public static double matgen(double[][] a, int lda, int n, double[] b) {
        double norma = 0.0;
        Random random = new Random(1325);

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                a[j][i] = random.nextDouble() - 0.5;
                if (a[j][i] > norma) {
                    norma = a[j][i];
                }
            }
        }
        Arrays.fill(b, 0.0);
        for (int j = 0; j < n; j++) {
            for (int i = 0; i < n; i++) {
                b[i] += a[j][i];
            }
        }
        return norma;
    }

    public static int dgefa(double[][] a, int lda, int n, int[] ipvt) {
        int info = 0;
        int nm1 = n - 1;

        if (nm1 >= 0) {
            for (int k = 0; k < nm1; k++) {
                int kp1 = k + 1;
                int l = idamax(n - k, a[k], k, 1) + k;
                ipvt[k] = l;

                if (a[k][l] != 0) {
                    if (l != k) {
                        double t = a[k][l];
                        a[k][l] = a[k][k];
                        a[k][k] = t;
                    }

                    double t = -1.0 / a[k][k];
                    dscal(n - kp1, t, a[k], kp1, 1);

                    for (int j = kp1; j < n; j++) {
                        double t2 = a[j][l];
                        if (l != k) {
                            a[j][l] = a[j][k];
                            a[j][k] = t2;
                        }
                        daxpy(n - kp1, t2, a[k], kp1, 1, a[j], kp1, 1);
                    }
                } else {
                    info = k;
                }
            }
        }
        ipvt[n - 1] = n - 1;
        if (a[n - 1][n - 1] == 0) {
            info = n - 1;
        }
        return info;
    }

    public static void dgesl(double[][] a, int lda, int n, int[] ipvt, double[] b, int job) {
        int nm1 = n - 1;

        if (job == 0) {
            if (nm1 >= 1) {
                for (int k = 0; k < nm1; k++) {
                    int l = ipvt[k];
                    double t = b[l];
                    if (l != k) {
                        b[l] = b[k];
                        b[k] = t;
                    }
                    daxpy(n - (k + 1), t, a[k], k + 1, 1, b, k + 1, 1);
                }
            }

            for (int kb = 0; kb < n; kb++) {
                int k = n - (kb + 1);
                b[k] /= a[k][k];
                double t = -b[k];
                daxpy(k, t, a[k], 0, 1, b, 0, 1);
            }
        } else {
            for (int k = 0; k < n; k++) {
                double t = ddot(k, a[k], 0, 1, b, 0, 1);
                b[k] = (b[k] - t) / a[k][k];
            }

            if (nm1 >= 1) {
                for (int kb = 1; kb < nm1; kb++) {
                    int k = n - (kb + 1);
                    int kp1 = k + 1;
                    b[k] += ddot(n - kp1, a[k], kp1, 1, b, kp1, 1);
                    int l = ipvt[k];
                    if (l != k) {
                        double t = b[l];
                        b[l] = b[k];
                        b[k] = t;
                    }
                }
            }
        }
    }

    public static void daxpy(int n, double da, double[] dx, int dxOff, int incx, double[] dy, int dyOff, int incy) {
        int ix = 0, iy = 0;

        if (n > 0 && da != 0) {
            if (incx != 1 || incy != 1) {
                if (incx < 0) ix = (-n + 1) * incx;
                if (incy < 0) iy = (-n + 1) * incy;

                for (int i = 0; i < n; i++) {
                    dy[iy + dyOff] += da * dx[ix + dxOff];
                    ix += incx;
                    iy += incy;
                }
            } else {
                for (int i = 0; i < n; i++) {
                    dy[i + dyOff] += da * dx[i + dxOff];
                }
            }
        }
    }

    public static double ddot(int n, double[] dx, int dxOff, int incx, double[] dy, int dyOff, int incy) {
        double dtemp = 0.0;
        int ix = 0, iy = 0;

        if (n > 0) {
            if (incx != 1 || incy != 1) {
                if (incx < 0) ix = (-n + 1) * incx;
                if (incy < 0) iy = (-n + 1) * incy;

                for (int i = 0; i < n; i++) {
                    dtemp += dx[ix + dxOff] * dy[iy + dyOff];
                    ix += incx;
                    iy += incy;
                }
            } else {
                for (int i = 0; i < n; i++) {
                    dtemp += dx[i + dxOff] * dy[i + dyOff];
                }
            }
        }
        return dtemp;
    }

    public static void dscal(int n, double da, double[] dx, int dxOff, int incx) {
        int nincx;

        if (n > 0) {
            if (incx != 1) {
                nincx = n * incx;
                for (int i = 0; i < nincx; i += incx) {
                    dx[i + dxOff] *= da;
                }
            } else {
                for (int i = 0; i < n; i++) {
                    dx[i + dxOff] *= da;
                }
            }
        }
    }

    public static int idamax(int n, double[] dx, int dxOff, int incx) {
        double dmax = 0.0, dtemp;
        int itemp = 0, ix = 0;

        if (n < 1) {
            itemp = -1;
        } else if (n == 1) {
            itemp = 0;
        } else if (incx != 1) {
            dmax = Math.abs(dx[0 + dxOff]);
            ix = 1 + incx;
            for (int i = 1; i < n; i++) {
                dtemp = Math.abs(dx[ix + dxOff]);
                if (dtemp > dmax) {
                    itemp = i;
                    dmax = dtemp;
                }
                ix += incx;
            }
        } else {
            dmax = Math.abs(dx[0 + dxOff]);
            for (int i = 1; i < n; i++) {
                dtemp = Math.abs(dx[i + dxOff]);
                if (dtemp > dmax) {
                    itemp = i;
                    dmax = dtemp;
                }
            }
        }
        return itemp;
    }

    public static double epslon(double x) {
        double a = 4.0 / 3.0;
        double eps = 0.0;

        while (eps == 0) {
            double b = a - 1.0;
            double c = b + b + b;
            eps = Math.abs(c - 1.0);
        }
        return eps * Math.abs(x);
    }

    public static void dmxpy(int n1, double[] y, int n2, int ldm, double[] x, double[][] m) {
        for (int j = 0; j < n2; j++) {
            for (int i = 0; i < n1; i++) {
                y[i] += x[j] * m[j][i];
            }
        }
    }

    public static LinpackResult linpack(int arraySize) {
        double mflopsResult = 0.0;
        double residnResult = 0.0;
        double timeResult = 0.0;
        double epsResult = 0.0;

        double[][] a = new double[arraySize][arraySize];
        double[] b = new double[arraySize];
        double[] x = new double[arraySize];
        double ops, total, norma, normx = 0.0;
        double resid, time;
        int n, lda = 0;
        int[] ipvt = new int[arraySize];

        lda = arraySize;
        n = arraySize;
        double nf = (double) n;

        ops = ((2.0 * nf) * nf * nf) / 3.0 + 2.0 * (nf * nf);

        norma = matgen(a, lda, n, b);
        time = (double) second() / 1000.0;
        dgefa(a, lda, n, ipvt);
        dgesl(a, lda, n, ipvt, b, 0);
        total = (double) second() / 1000.0 - time;

        System.arraycopy(b, 0, x, 0, n);
        norma = matgen(a, lda, n, b);
        for (int i = 0; i < n; i++) {
            b[i] = -b[i];
        }
        dmxpy(n, b, n, lda, x, a);
        resid = 0.0;
        normx = 0.0;
        for (int i = 0; i < n; i++) {
            resid = Math.max(resid, Math.abs(b[i]));
            normx = Math.max(normx, Math.abs(x[i]));
        }

        epsResult = epslon(1.0);

        residnResult = resid / (n * norma * normx * epsResult);
        residnResult = Math.round(residnResult * 100.0) / 100.0;

        timeResult = Math.round(total * 100.0) / 100.0;

        mflopsResult = ops / (1.0e6 * total);
        mflopsResult = Math.round(mflopsResult * 1000.0) / 1000.0;

        LinpackResult result = new LinpackResult();
        result.norma = norma;
        result.residual = resid;
        result.normalisedResidual = residnResult;
        result.epsilon = epsResult;
        result.time = timeResult;
        result.mflops = mflopsResult;

        System.out.println(result);

        return result;
    }

    public static void main(String[] args) {
        long startTimeMs = System.currentTimeMillis();

        linpack(ARRAY_SIZE);

        long endTimeMs = System.currentTimeMillis();
        long durationMs = endTimeMs - startTimeMs;

        System.out.printf("Execution time: %dms\n", durationMs);
    }
}

File: langs/rust/linpack/Linpack.rs
use std::time::{SystemTime, UNIX_EPOCH};

const ARRAY_SIZE: usize = 2000;

#[derive(Debug)]
#[allow(dead_code)]
struct LinpackResult {
    norma: f64,
    residual: f64,
    normalised_residual: f64,
    epsilon: f64,
    time: f64,
    mflops: f64,
}

fn second() -> i64 {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap();
    now.as_millis() as i64
}

fn lran(seed: &mut [i32; 4]) -> f64 {
    let m1 = 494;
    let m2 = 322;
    let m3 = 2508;
    let m4 = 2549;
    let ipw2 = 4096;

    let r = 1.0 / ipw2 as f64;

    let mut it4 = seed[3] * m4;
    let mut it3 = it4 / ipw2;
    it4 -= ipw2 * it3;
    it3 += seed[2] * m4 + seed[3] * m3;
    let mut it2 = it3 / ipw2;
    it3 -= ipw2 * it2;
    it2 += seed[1] * m4 + seed[2] * m3 + seed[3] * m2;
    let mut it1 = it2 / ipw2;
    it2 -= ipw2 * it1;
    it1 += seed[0] * m4 + seed[1] * m3 + seed[2] * m2 + seed[3] * m1;
    it1 %= ipw2;

    seed[0] = it1;
    seed[1] = it2;
    seed[2] = it3;
    seed[3] = it4;

    r * (it1 as f64 + r * (it2 as f64 + r * (it3 as f64 + r * it4 as f64)))
}

fn matgen(a: &mut Vec<Vec<f64>>, _lda: usize, n: usize, b: &mut Vec<f64>) -> f64 {
    let mut norma = 0.0;
    let mut iseed = [1, 2, 3, 1325];

    for i in 0..n {
        for j in 0..n {
            a[j][i] = lran(&mut iseed) - 0.5;
            if a[j][i] > norma {
                norma = a[j][i];
            }
        }
    }

    for i in 0..n {
        b[i] = 0.0;
    }

    for j in 0..n {
        for i in 0..n {
            b[i] += a[j][i];
        }
    }

    norma
}

fn daxpy(n: usize, da: f64, dx: &[f64], dx_off: usize, incx: usize, dy: &mut [f64], dy_off: usize, incy: usize) {
    if n > 0 && da != 0.0 {
        if incx != 1 || incy != 1 {
            let mut ix: isize = 0;
            let mut iy: isize = 0;
            let n = n as isize;
            let incx = incx as isize;
            let incy = incy as isize;
            if incx < 0 {
                ix = (-n + 1) * incx;
            }
            if incy < 0 {
                iy = (-n + 1) * incy;
            }
            for _ in 0..n {
                dy[(iy as usize) + dy_off] += da * dx[(ix as usize) + dx_off];
                ix += incx;
                iy += incy;
            }
        } else {
            for i in 0..n {
                dy[i + dy_off] += da * dx[i + dx_off];
            }
        }
    }
}

fn ddot(n: usize, dx: &[f64], dx_off: usize, incx: usize, dy: &[f64], dy_off: usize, incy: usize) -> f64 {
    let mut dtemp = 0.0;
    if n > 0 {
        if incx != 1 || incy != 1 {
            let mut ix: isize = 0;
            let mut iy: isize = 0;
            let n = n as isize;
            let incx = incx as isize;
            let incy = incy as isize;
            if incx < 0 {
                ix = (-n + 1) * incx;
            }
            if incy < 0 {
                iy = (-n + 1) * incy;
            }
            for _ in 0..n {
                dtemp += dx[(ix as usize) + dx_off] * dy[(iy as usize) + dy_off];
                ix += incx;
                iy += incy;
            }
        } else {
            for i in 0..n {
                dtemp += dx[i + dx_off] * dy[i + dy_off];
            }
        }
    }
    dtemp
}

fn dscal(n: usize, da: f64, dx: &mut [f64], dx_off: usize, incx: usize) {
    if n > 0 {
        if incx != 1 {
            let nincx = n * incx;
            let mut i = 0;
            while i < nincx {
                dx[i + dx_off] *= da;
                i += incx;
            }
        } else {
            for i in 0..n {
                dx[i + dx_off] *= da;
            }
        }
    }
}

fn idamax(n: usize, dx: &[f64], dx_off: usize, incx: usize) -> usize {
    let mut itemp = 0usize;
    if n < 1 {
        itemp = usize::MAX; // will be ignored
    } else if n == 1 {
        itemp = 0;
    } else if incx != 1 {
        let mut dmax = dx[0 + dx_off].abs();
        let mut ix = 1 + incx;
        for i in 1..n {
            let dtemp = dx[ix + dx_off].abs();
            if dtemp > dmax {
                itemp = i;
                dmax = dtemp;
            }
            ix += incx;
        }
    } else {
        itemp = 0;
        let mut dmax = dx[0 + dx_off].abs();
        for i in 1..n {
            let dtemp = dx[i + dx_off].abs();
            if dtemp > dmax {
                itemp = i;
                dmax = dtemp;
            }
        }
    }
    itemp
}

fn epslon(x: f64) -> f64 {
    let a = 4.0f64 / 3.0f64;
    let mut eps = 0.0f64;
    while eps == 0.0 {
        let b = a - 1.0f64;
        let c = b + b + b;
        eps = (c - 1.0f64).abs();
    }
    eps * x.abs()
}

fn dmxpy(n1: usize, y: &mut [f64], n2: usize, _ldm: usize, x: &[f64], m: &Vec<Vec<f64>>) {
    for j in 0..n2 {
        for i in 0..n1 {
            y[i] += x[j] * m[j][i];
        }
    }
}

#[allow(unused_comparisons)]
fn dgefa(a: &mut Vec<Vec<f64>>, _lda: usize, n: usize, ipvt: &mut [usize]) -> usize {
    let mut info = 0usize;
    let nm1 = n - 1;
    if nm1 >= 0 {
        for k in 0..nm1 {
            let kp1 = k + 1;
            let l = idamax(n - k, &a[k], k, 1) + k;
            ipvt[k] = l;
            if a[k][l] != 0.0 {
                if l != k {
                    let t = a[k][l];
                    a[k][l] = a[k][k];
                    a[k][k] = t;
                }
                let t = -1.0 / a[k][k];
                dscal(n - kp1, t, &mut a[k], kp1, 1);
                for j in kp1..n {
                    let tcol = a[j][l];
                    if l != k {
                        let temp = a[j][k];
                        a[j][k] = tcol;
                        a[j][l] = temp;
                    }
                    for i in kp1..n {
                        a[j][i] += tcol * a[k][i];
                    }
                }
            } else {
                info = k;
            }
        }
    }
    ipvt[n - 1] = n - 1;
    if a[n - 1][n - 1] == 0.0 {
        info = n - 1;
    }
    info
}

fn dgesl(a: &Vec<Vec<f64>>, _lda: usize, n: usize, ipvt: &[usize], b: &mut [f64], job: i32) {
    let nm1 = n - 1;
    if job == 0 {
        if nm1 >= 1 {
            for k in 0..nm1 {
                let l = ipvt[k];
                let t = b[l];
                if l != k {
                    b[l] = b[k];
                    b[k] = t;
                }
                let kp1 = k + 1;
                daxpy(n - kp1, t, &a[k], kp1, 1, b, kp1, 1);
            }
        }
        for kb in 0..n {
            let k = n - (kb + 1);
            b[k] /= a[k][k];
            let t = -b[k];
            daxpy(k, t, &a[k], 0, 1, b, 0, 1);
        }
    } else {
        for k in 0..n {
            let t = ddot(k, &a[k], 0, 1, b, 0, 1);
            b[k] = (b[k] - t) / a[k][k];
        }
        if nm1 >= 1 {
            for kb in 1..nm1 {
                let k = n - (kb + 1);
                let kp1 = k + 1;
                b[k] += ddot(n - kp1, &a[k], kp1, 1, b, kp1, 1);
                let l = ipvt[k];
                if l != k {
                    let t = b[l];
                    b[l] = b[k];
                    b[k] = t;
                }
            }
        }
    }
}

fn linpack(array_size: usize) -> LinpackResult {
    let mut a = vec![vec![0.0_f64; array_size]; array_size];
    let mut b = vec![0.0_f64; array_size];
    let mut x = vec![0.0_f64; array_size];
    let mut ipvt = vec![0usize; array_size];

    let nf = array_size as f64;
    let ops = (2.0 * nf * nf * nf) / 3.0 + 2.0 * (nf * nf);

    matgen(&mut a, array_size, array_size, &mut b);
    let norma;
    let time = second() as f64 / 1000.0;
    dgefa(&mut a, array_size, array_size, &mut ipvt);
    dgesl(&a, array_size, array_size, &ipvt, &mut b, 0);
    let total = second() as f64 / 1000.0 - time;

    for i in 0..array_size {
        x[i] = b[i];
    }
    norma = matgen(&mut a, array_size, array_size, &mut b);
    for i in 0..array_size {
        b[i] = -b[i];
    }
    dmxpy(array_size, &mut b, array_size, array_size, &x, &a);

    let mut resid = 0.0;
    let mut normx = 0.0;
    for i in 0..array_size {
        if b[i].abs() > resid {
            resid = b[i].abs();
        }
        if x[i].abs() > normx {
            normx = x[i].abs();
        }
    }

    let eps_result = epslon(1.0);
    let mut residn_result = resid / (array_size as f64 * norma * normx * eps_result);
    residn_result += 0.005;
    residn_result = (residn_result * 100.0).floor();
    residn_result /= 100.0;

    let mut time_result = total;
    time_result += 0.005;
    time_result = (time_result * 100.0).floor();
    time_result /= 100.0;

    let mut mflops_result = ops / (1.0e6 * total);
    mflops_result += 0.0005;
    mflops_result = (mflops_result * 1000.0).floor();
    mflops_result /= 1000.0;

    let result = LinpackResult {
        norma,
        residual: resid,
        normalised_residual: residn_result,
        epsilon: eps_result,
        time: time_result,
        mflops: mflops_result,
    };

    result
}

fn main() {
    let start = std::time::Instant::now();

    let result = linpack(ARRAY_SIZE);
    println!("{:?}", result);

    let elapsed = start.elapsed();
    println!("Execution time: {}ms", elapsed.as_millis());
}

